designer/server/src/main/scala/pl/touk/nussknacker/ui/process/migrate/RemoteEnvironment.scala

```diff
@@ -195,25 +216,73 @@ trait StandardRemoteEnvironment extends FailFastCirceSupport with RemoteEnvironm
       processName: ProcessName,
       isFragment: Boolean
   )(implicit ec: ExecutionContext, loggedUser: LoggedUser): Future[Either[NuDesignerError, Unit]] = {
-    val migrateScenarioRequest: MigrateScenarioRequest =
-      MigrateScenarioRequest(
-        environmentId,
-        processingMode,
-        engineSetupName,
-        processCategory,
-        scenarioGraph,
-        processName,
-        isFragment
+
+    val result: EitherT[Future, NuDesignerError, Unit] = for {
+      remoteScenarioDescriptionVersion <- fetchRemoteMigrationScenarioDescriptionVersion
+      localScenarioDescriptionVersion = migrationApiAdapterService.getCurrentApiVersion
+      migrateScenarioRequest: MigrateScenarioData =
+        MigrateScenarioDataV2(
+          environmentId,
+          loggedUser.username,
+          processingMode,
+          engineSetupName,
+          processCategory,
+          scenarioGraph,
+          processName,
+          isFragment
+        )
+      versionsDifference = localScenarioDescriptionVersion - remoteScenarioDescriptionVersion
+      transformedMigrateScenarioRequestE =
+        if (versionsDifference > 0)
+          migrationApiAdapterService.adaptDown(migrateScenarioRequest, versionsDifference)
```

Comment:
WDYT about changing the interface a little bit? IMO we could look like this:
```
def adaptTo(data: D, targetDataVersion: Int): Either[ApiAdapterServiceError, D] =
```

It seems we can encapsulate more adapting logic inside the adapter